src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -54,11 +107,35 @@ typedef struct ucc_tl_mlx5_mcast_context_config {
 } ucc_tl_mlx5_mcast_context_config_t;
 
 typedef struct ucc_tl_mlx5_mcast_lib {
+    ucc_tl_lib_t                   super;
+    ucc_tl_mlx5_mcast_lib_config_t cfg;
 } ucc_tl_mlx5_mcast_lib_t;
 UCC_CLASS_DECLARE(ucc_tl_mlx5_mcast_lib_t, const ucc_base_lib_params_t *,
                   const ucc_base_config_t *);
 
-typedef struct ucc_tl_mlx5_mcast_ctx_params {
+typedef struct ucc_tl_mlx5_mcast_oob_ctx {
+    void               *ctx;
+    union {
+        ucc_oob_coll_t *oob;
+        ucc_subset_t    subset;
+    };
+} ucc_tl_mlx5_mcast_oob_ctx_t;
+
+typedef struct ucc_tl_mlx5_mcast_reg {
+    void *mr;
+} ucc_tl_mlx5_mcast_reg_t;
+
+typedef struct ucc_tl_mlx5_mcast_rcache_region {
+    ucc_rcache_region_t     super;
+    ucc_tl_mlx5_mcast_reg_t reg;
+} ucc_tl_mlx5_mcast_rcache_region_t;
+
+typedef struct mcast_ctx_params {
+    char    *ib_dev_name;
+    int      print_nack_stats;
+    int      timeout;
+    int      (*bcast) (void*, void*, int, int);
```

Comment:
Can you explain why these function belong to `ctx_params`? How can they be runtime configurable parameters, and why should we make them configurable? Do they really belong to the context and not the team ?

Reply:
Because we would like the context to have the flexibility to choose any out of band bcast() function for its context initialization purposes. I chose it to be under parameters as it is an input parameter that is fed into ucc_tl_mlx5_mcast_context_init() function.

Reply:
Let me clarify my concerns:
- any tl "context_config" struct (here `ucc_tl_mlx5_context_config`) contains fields that are configurable at runtime through environment variables. This is why I think that bcast doesn't belong there
- I think I remember from the full code that bcast is actually used at team initialization. If I am correct, this means that 1) maybe it doesn't belong to ctx configuration, and 2) since team creation is not blocking, we would actually need a bcast_post and bcast_test

Reply:
Thanks for the clarification. Since right now, we are not reading those functions from env var, then I am going to remove them from here. 
_also I see that appears both in ucc_tl_mlx5_context_config and in ucc_tl_mlx5_mcast_coll_context. Is it a duplicate?_ It is indeed a duplicate, we copy those configurations to ucc_tl_mlx5_mcast_coll_context for ease of access and to avoid dealing with mlx5 structs under the hood in mcast.